File: cpp/src/standard_library/standard_math_library.cpp
#include "standard_math_library.hpp"

#include <math.h>

#include "../virtual_machine.hpp"
#include "../values/object_value.hpp"
#include "../utils.hpp"
#include "../scope.hpp"

#define M_DEG_TO_RAD M_PI / 180.0

namespace lysithea_vm
{
    std::shared_ptr<const scope> standard_math_library::library_scope = create_scope();

    std::shared_ptr<scope> standard_math_library::create_scope()
    {
        auto result = std::make_shared<scope>();

        auto functions = std::make_shared<object_value>();

        functions->data["E"] = value(M_E);
        functions->data["PI"] = value(M_PI);
        functions->data["DegToRad"] = value(M_DEG_TO_RAD);

        functions->data["sin"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &top = args.get_number(0);
            vm.push_stack(sin(top));
        });
        functions->data["cos"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &top = args.get_number(0);
            vm.push_stack(cos(top));
        });
        functions->data["tan"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &top = args.get_number(0);
            vm.push_stack(tan(top));
        });

        functions->data["pow"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            const auto &y = args.get_number(1);
            vm.push_stack(pow(x, y));
        });
        functions->data["exp"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(exp(x));
        });
        functions->data["floor"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(floor(x));
        });
        functions->data["ceil"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(ceil(x));
        });
        functions->data["round"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(round(x));
        });
        functions->data["isNaN"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(std::isnan(x));
        });
        functions->data["isFinite"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(std::isfinite(x));
        });
        functions->data["parse"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &top = args.get_index(0);
            if (top.is_number())
            {
                vm.push_stack(top);
                return;
            }

            vm.push_stack(std::stod(top.to_string()));
        });

        functions->data["log"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(log(x));
        });
        functions->data["log2"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(log2(x));
        });
        functions->data["log10"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(log10(x));
        });
        functions->data["abs"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            const auto &x = args.get_number(0);
            vm.push_stack(abs(x));
        });

        functions->data["max"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            auto max = args.get_index(0);
            for (auto iter = args.data.cbegin() + 1; iter != args.data.cend(); ++iter)
            {
                if (iter->compare_to(max) > 0)
                {
                    max = *iter;
                }
            }

            vm.push_stack(max);
        });

        functions->data["min"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            auto min = args.get_index(0);
            for (auto iter = args.data.cbegin() + 1; iter != args.data.cend(); ++iter)
            {
                if (iter->compare_to(min) < 0)
                {
                    min = *iter;
                }
            }

            vm.push_stack(min);
        });

        functions->data["sum"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            auto total = 0.0;
            for (const auto &iter : args.data)
            {
                if (!iter.is_number())
                {
                    throw std::runtime_error("Invalid addition operator usage, must be all numbers");
                }
                total += iter.get_number();
            }
            vm.push_stack(total);
        });

        return result;
    }
} // lysithea_vm

File: ts/src/standardLibrary/standardMathLibrary.ts
import { Scope, IReadOnlyScope } from "../scope";
import { BuiltinFunctionValue } from "../values/builtinFunctionValue";
import { NumberValue, isNumberValue } from "../values/numberValue";
import { ObjectValue, ObjectValueMap } from "../values/objectValue";

const degToRad = Math.PI / 180.0;

export const mathScope: IReadOnlyScope = createMathScope();

export function createMathScope()
{
    const result = new Scope();

    const mathFunctions: ObjectValueMap =
    {
        E: new NumberValue(Math.E),
        PI: new NumberValue(Math.PI),
        DegToRad: new NumberValue(degToRad),

        sin: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackNumber(Math.sin(args.getNumber(0)));
        }, "math.sin"),
        cos: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackNumber(Math.cos(args.getNumber(0)));
        }, "math.cos"),
        tan: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackNumber(Math.tan(args.getNumber(0)));
        }, "math.tan"),
        exp: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackNumber(Math.exp(args.getNumber(0)));
        }, "math.exp"),
        ceil: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackNumber(Math.ceil(args.getNumber(0)));
        }, "math.ceil"),
        floor: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackNumber(Math.floor(args.getNumber(0)));
        }, "math.floor"),
        round: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackNumber(Math.round(args.getNumber(0)));
        }, "math.round"),
        isFinite: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackBool(Number.isFinite(args.getNumber(0)));
        }, "math.isFinite"),
        isNaN: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackBool(isNaN(args.getNumber(0)));
        }, "math.isNaN"),
        parse: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndex(0);
            if (isNumberValue(top))
            {
                vm.pushStack(top);
            }
            else
            {
                vm.pushStackNumber(parseFloat(top.toString()));
            }
        }, "math.parse"),
        log: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackNumber(Math.log(args.getNumber(0)));
        }, "math.log"),
        abs: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStackNumber(Math.abs(args.getNumber(0)));
        }, "math.abs"),
        max: new BuiltinFunctionValue((vm, args) =>
        {
            let max = args.value[0];
            for (let i = 1; i < args.value.length; i++)
            {
                const next = args.value[i];
                if (next.compareTo(max) > 0)
                {
                    max = next;
                }
            }
            vm.pushStack(max);
        }, "math.max"),
        min: new BuiltinFunctionValue((vm, args) =>
        {
            let min = args.value[0];
            for (let i = 1; i < args.value.length; i++)
            {
                const next = args.value[i];
                if (next.compareTo(min) < 0)
                {
                    min = next;
                }
            }
            vm.pushStack(min);
        }, "math.min"),
        sum: new BuiltinFunctionValue((vm, args) =>
        {
            let result = 0.0;
            for (let i = 0; i < args.value.length; i++)
            {
                const item = args.value[i];
                if (isNumberValue(item))
                {
                    result += item.value;
                }
                else
                {
                    throw new Error('Addition operator expects all numbers');
                }
            }
            vm.pushStackNumber(result);
        }, "math.sum")
    };

    result.define('math', new ObjectValue(mathFunctions));

    return result;
}
